### PR TITLE
Hdmi cec control

### PR DESCRIPTION
**Pull request by danielandersson**
Implementation for #365 

TODO:

* [ ] Add some automated tests  – only been tested in practice.
* [ ] Need to workaround the CEC-addressing.
* [ ] Make source and destination addresses configurable

### CODE REVIEW

File: _stbt/control.py [hunk 1]
```diff
@@ -42,7 +42,10 @@ def uri_to_remote(uri, display=None):
         (r'x11:(?P<display>[^,]+)?(,(?P<mapping>.+)?)?', _X11Remote),
     ]
     if gpl_controls is not None:
+        print "GPL Controls ACTIVATED"
         remotes += gpl_controls
+
+    print(remotes)
```
Comment (wmanley) [inline _stbt/control.py]: You'll want to remove these debug statements.

File: _stbt/control_gpl.py [hunk 2]
```diff
@@ -1,16 +1,68 @@
+from .logging import debug
+
+
 class HdmiCecControl(object):
+    cecconfig = {}
+    lib = {}
```
Comment (wmanley) [inline _stbt/control_gpl.py]: I don't think either of these should be here.  They look like they should be attached to the object rather than the class.  Removing these two lines should be sufficient.

File: _stbt/control_gpl.py [hunk 3]
```diff
@@ -1,16 +1,68 @@
...
+        self.cecconfig.strDeviceName = "STB-Tester"
+        self.cecconfig.bActivateSource = 0
+        self.cecconfig.deviceTypes.Add(cec.CEC_DEVICE_TYPE_RECORDING_DEVICE)
+        self.cecconfig.clientVersion = cec.LIBCEC_VERSION_CURRENT
+        self.lib = cec.ICECAdapter.Create(self.cecconfig)
+        # print libCEC version and compilation information
+        debug("libCEC version " + self.lib.VersionToString(self.cecconfig.serverVersion) +
+              " loaded: " + self.lib.GetLibInfo())
+
+        if device == 'auto':
```
Comment (wmanley) [inline _stbt/control_gpl.py]: I would suggest making this conditional `device is None`.  We can change the regex below to pass `device=None` if you specify `hdmi-cec:` as your control.  This will be more consistent with the other remotes.

File: _stbt/control_gpl.py [hunk 4]
```diff
@@ -1,16 +1,68 @@
...
+            device = self.DetectAdapter()
+            if device is None:
+                debug("No adapters found")
```
Comment (wmanley) [inline _stbt/control_gpl.py]: This should probably be raising an exception rather than printing to debug.  Perhaps create `HdmiCecException` with:

    class HdmiCecError(Exception):
        pass

and then do `raise HdmiCecError("No adaptors found")` here.

File: _stbt/control_gpl.py [hunk 5]
```diff
@@ -1,16 +1,68 @@
...
+        if self.lib.Open(device):
+            debug("connection opened")
+        else:
+            debug("failed to open a connection to the CEC adapter")
```
Comment (wmanley) [inline _stbt/control_gpl.py]: This should also be raising an exception.

File: _stbt/control_gpl.py [hunk 6]
```diff
@@ -1,16 +1,68 @@
...
+
+    # Map our recommended keynames (from linux input-event-codes.h) to the
+    # equivalent CEC commands.
+    _KEYNAMES = {
+        "KEY_OK": "14:44:00",
+        "KEY_UP": "14:44:01",
+        "KEY_DOWN": "14:44:02",
+        "KEY_LEFT": "14:44:03",
+        "KEY_RIGHT": "14:44:04",
+        "KEY_BACK": "14:44:0D"
```
Comment (wmanley) [inline _stbt/control_gpl.py]: Do you have a reference for these codes here?  It would be good to link to it in the comment you've made above.

File: _stbt/control_gpl.py [hunk 7]
```diff
@@ -1,16 +1,68 @@
...
+                debug("No adapters found")
+        if self.lib.Open(device):
+            debug("connection opened")
+        else:
+            debug("failed to open a connection to the CEC adapter")
 
     def press(self, key):
-        import cec
-        raise NotImplementedError()
+        cec_command = self._KEYNAMES.get(key, key)
```
Comment (wmanley) [inline _stbt/control_gpl.py]: Right, so key can be either one of our symbolic keys like `KEY_UP` or a CEC command of the form `14:xx`? What use-case had you in mind for the latter?

File: _stbt/control_gpl.py [hunk 8]
```diff
@@ -1,16 +1,68 @@
...
+        cmd = self.lib.CommandFromString(cec_command)
+        debug("transmit " + key + " as " + cec_command)
+        if self.lib.Transmit(cmd):
+            self.lib.Transmit(self.lib.CommandFromString('14:45'))
+            debug("command sent")
+        else:
+            debug("failed to send command")
```
Comment (wmanley) [inline _stbt/control_gpl.py]: This should probably be `raise HdmiCecError("failed to send command")`.

File: _stbt/control_gpl.py [hunk 9]
```diff
@@ -1,16 +1,68 @@
...
+
+    # detect an adapter and return the com port path
+    def DetectAdapter(self):
```
Comment (wmanley) [inline _stbt/control_gpl.py]: Could you rename this `detect_adapter` for [pep8 compliance](https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables)?

File: _stbt/control_gpl.py [hunk 10]
```diff
@@ -1,16 +1,68 @@
...
+        retval = None
+        adapters = self.lib.DetectAdapters()
+        for adapter in adapters:
+            debug("found a CEC adapter:")
+            debug("port:     " + adapter.strComName)
+            debug("vendor:   " + hex(adapter.iVendorId))
+            debug("product:  " + hex(adapter.iProductId))
+            retval = adapter.strComName
```
Comment (wmanley) [inline _stbt/control_gpl.py]: This method will return the last adapter that `lib.DetectAdapters` lists.  Is this intentional?  Perhaps this should just be `return adapter.strComName` unless I've misunderstood?

**Comment by danielandersson** [comment]
Hey @wmanley, 
I've adjusted the code according to your comments. Most of it was straightforward and leftovers from the pyCecClient that was the inspiration. 

Great comment with the new exception.
For the CEC codes, I added a short description and a link to a extremely nice site to lookup the codes for a specific command. Moreover I think a Wiki-page with a wider description of CEC might be necessary.
As for the "lib.DetectAdapters" you're absolutely correct is lists all connected adapters. In the above mentioned [pyCeceClient](https://github.com/Pulse-Eight/libcec/blob/master/src/pyCecClient/pyCecClient.py) they uses it like this. Thats is the reason I left it. If you're having multiple adapters you have to specify the correct one in _stbt.conf_.

Do you have any suggestions on how to solve the CEC addressing issue? The first two numbers in the CEC-codes, in all of the examples in the code now - 14, is first the source and secondly the destination. These two number have to be configured, at least the destination, by the end-user. The problem is that it isn't that user-friendly to identify what ID the device-under-testing have. Any ideas?

**Comment by wmanley** [comment]
> Do you have any suggestions on how to solve the CEC addressing issue? The first two numbers in the CEC-codes, in all of the examples in the code now - 14, is first the source and secondly the destination.

I was having a look at the excellent CEC website you shared.  What about source: TV, destination: Broadcast (0F).  Would that work?  I think we'd also want to make either option configurable.  I'll make a comment in-line describing how to do this.

File: _stbt/control_gpl.py [hunk 11]
```diff
@@ -1,16 +1,72 @@
...
+    _KEYNAMES = {
+        "KEY_OK": "14:44:00",
+        "KEY_UP": "14:44:01",
+        "KEY_DOWN": "14:44:02",
+        "KEY_LEFT": "14:44:03",
+        "KEY_RIGHT": "14:44:04",
+        "KEY_BACK": "14:44:0D"
+    }
+
     def __init__(self, device):
```
Comment (wmanley) [inline _stbt/control_gpl.py]: To make the source and destinations configurable make this

    def __init__(self, device=None, source=None, destination=None):

and add

    if source is None:
        source = 0
    if destination is None:
        destination = 0xf

here

File: _stbt/control_gpl.py [hunk 12]
```diff
@@ -1,16 +1,72 @@
...
 
 
 def test_hdmi_cec_control():
-    from .control import uri_to_remote
-    control = uri_to_remote('hdmi-cec:192.168.1.3')
+    # from .control import uri_to_remote
+    # control = uri_to_remote('hdmi-cec:192.168.1.3')
     assert False, "Tests not written"
 
 controls = [
```
Comment (wmanley) [inline _stbt/control_gpl.py]: and modify the regex here to say:

     (r'hdmi-cec:(?P<device>[^:]+)(:(?P<source>[^:]+)(:(?P<destination>[^:]+))?)?', HdmiCecControl),

(not tested)

**Comment by wmanley** [comment]
I think you'll need to open a new pull request too.  Above it says:

> @danielandersson wants to merge 4 commits into `stb-tester:hdmi-cec-control` from `danielandersson:hdmi-cec-control`

Whereas it should say:

> @danielandersson wants to merge 4 commits into `stb-tester:master` from `danielandersson:hdmi-cec-control`

(merging into stb-tester:master).

I don't know how to change this without closing this pull-request and opening a new one.

**Comment by danielandersson** [comment]
Wouldn't it be more suitable to have "Recording 1" as default source since it describes the device best? If we decide to use TV it might collide in case a TV is connected and I know from my experiments that for example a Playstation 4 depends on having a device with CEC+EDID to control it, something TV usually have but not all USB-CEC adapters.

As for the destination, having broadcast as default feels like a dirty solution but also the most user-friendly one. Not optimal if multiple devices are connected to the CEC chain.

**Comment by wmanley** [comment]
> Wouldn't it be more suitable to have "Recording 1" as default source since it describes the device best

I don't think so.  stb-tester is intended to simulate actual usage of set-top boxes as if a human were using it.  Humans use them via TVs so stb-tester should pretend to be a TV.

> I know from my experiments that for example a Playstation 4 depends on having a device with CEC+EDID to control it, something TV usually have but not all USB-CEC adapters.

So you're saying that it doesn't work if the USB-CEC adapter pretends to be a TV, whereas it does work when identifying itself as "Recording 1"?

> As for the destination, having broadcast as default feels like a dirty solution but also the most user-friendly one. Not optimal if multiple devices are connected to the CEC chain.

I admit I know little about HDMI-CEC, but I would have thought that if you can have multiple devices in a CEC chain there must be some way that they avoid sender ID conflicts?  Do you know how these numbers get assigned in practice?  Could we query that somehow?

OTOH, if broadcast works for all the cases we care about then maybe it's not that big a deal?

**Comment by danielandersson** [comment]
I understand and agree that the "consuming side" of the HDMI, which is the capture card and USB-CEC adapter, should be considered the TV preferable. As you say, the purpose is to simulate and end-to-end usage using a remote to control and output image to validate. However, for the perspective of the HDMI signal and CEC – the capture card "side" records the signal.

I have experimented with Apple TV and Playstation 4 and for the first mentioned it worked as intended. The Apple TV requested a CEC-address (Playback 1) and the USB-CEC adapter requested Recorder 1. No problems.
When I tried with Playstation 4 however, it didn't request a CEC-address when only having the USB-CEC adapter and PS4 in the HDMI chain. The PS4 was totally quite and didn't listen nor say anything. After some investigation I found that when connecting the PS4 to a TV the CEC, using the TV remote, was fully working. To debug this I connected the USB-CEC adapter to the very same TV, having the adapter and PS4 connected at the same time.

                                                    TV
                                   TV HDMI 1  -------------  TV HDMI 2
                                  |                                   |
                                  |                                   |
                                  |                                   |
                                Playstation 4                   USB-CEC 

This allowed me to listen to the CEC channel and it was using this I discovered that the PS4 must be communicating using another channel before initing CEC.
I found after searching online that PS4 listens on the EDID channel before enabling CEC. And since the USB-CEC adapter does not have the EDID channel it can't enable CEC on PS4.

So with this issue with PS4, if we set the USB-CEC adapter as TV by default in STB-Tester and since there is only one slot for TV in the CEC-addressing (apart from the reserved address "E") we will occupy and it might lead to potential issues.
However, as the PS4 was connected to the TV the capture card couldn't capture the image so thing might be a corner case. (Since it is dependent on stbt camera)


TLDR;
> Humans use them via TVs so stb-tester should pretend to be a TV.

I agree but I think we should make an exception and have it as recording device to not interfere with TV.

> So you're saying that it doesn't work if the USB-CEC adapter pretends to be a TV ...

From my tries, it has worked both as TV and recorder.

> they avoid sender ID conflicts?

I'm do not understand this fully but from what I understand they mitigate the conflict by taking another slot. That is why you have multiple addresses for most of the types, "Recording 1 & Recording 2" etc.

`self.cecconfig.deviceTypes.Add(cec.CEC_DEVICE_TYPE_RECORDING_DEVICE)`

**Comment by SidePlace** [comment]
Hi,
Anyone still looking into this branch? Any chance we still having this? I really want the CEC support.

**Comment by drothlis** [comment]
Moved to #393.
